### PR TITLE
docker: Optimize driver container image size

### LINKED ISSUE #49: How to handle scenario where out-of-tree driver depends on in-tree driver(s)
### **Summary**
The Intel GPU i915 driver has a large dependency tree. It depends on several in-tree drivers present in /lib/modules/(KERNELRELEASE)/. In the scenario where an out of tree driver depends on in-tree driver(s), the user needs to copy in-tree drivers into /opt/lib/modules/(KERNELRELEASE)/ rather than leveraging the in-tree drivers directly on the host. This is because everything is done in /opt including modprobe -d /opt (done explicitly by KMM) as well as depmod -b /opt (done in dockerfile).

### Goal 
The goal is to determine if we can avoid copying in-tree drivers. Only out of tree drivers should be a part of the driver container image.

### Possible Solution
This is a complex problem where a solution may not feasible. If we continue to use /opt to avoid tainting the default /lib directory, it would require copying in-tree drivers from the host by using a modules.dep file (this file would contain dependency list and would be generated after placing the out of tree drivers modules in /lib). From a KMM perspective, it would receive a driver container with exclusively out of tree drivers and then use the supplied modules.dep to copy over the necessary in-tree drivers to successfully load the out of tree driver. This solution could involve host mounting and sym links.

### PR DESCRIPTION
**Pull request by hershpa**
Fixes [#49](https://github.com/intel/intel-data-center-gpu-driver-for-openshift/issues/49)
Copy only out-of-tree drivers to /opt/lib/modules/${KERNEL_FULL_VERSION}/ in driver container.
Create symlink from /opt to /lib to use in-tree kmods on the build node's filesystem.
Run depmod in the first stage DTK environment to generate dependency list.
At runtime, KMM follows symlink to load the in-tree dependencies.
See link for more details: https://kmm.sigs.k8s.io/documentation/kmod_image/#depending-on-in-tree-kernel-modules
Signed-off-by: Hersh Pathak hersh.pathak@intel.com

### CODE REVIEW

**Review by mregmi — APPROVED** [review]
LGTM

File: docker/intel-dgpu-driver.Dockerfile [hunk 1]
```diff
@@ -34,6 +34,15 @@ RUN git clone -b ${I915_RELEASE} --single-branch https://github.com/intel-gpu/in
     && cp defconfigs/i915 .config \
     && make olddefconfig && make modules -j $(nproc) && make modules_install
 
+# Copy out-of-tree drivers to /opt/lib/modules/${KERNEL_FULL_VERSION}/
+RUN for file in $(find /lib/modules/${KERNEL_FULL_VERSION}/updates/ -name "*.ko"); do \
+    dname=$(dirname /opt$file); mkdir -p $dname; cp $file /opt$file; done
```
Comment (uMartinXu) [inline docker/intel-dgpu-driver.Dockerfile]: Can you try command below? Make sure directory /opt already exists. 

```
RUN for file in $(find /lib/modules/${KERNEL_FULL_VERSION}/updates/ -name "*.ko"); do \
    cp $file /opt --parents; \
done
```

File: docker/intel-dgpu-driver.Dockerfile [hunk 2]
```diff
@@ -34,6 +34,15 @@ RUN git clone -b ${I915_RELEASE} --single-branch https://github.com/intel-gpu/in
     && cp defconfigs/i915 .config \
```
Comment (uMartinXu) [inline docker/intel-dgpu-driver.Dockerfile:34]: Can we add some links to describe the technical detail in the commit log?

Comment (uMartinXu) [reply docker/intel-dgpu-driver.Dockerfile:34]: Also the related github issue link should be added into the commit log.

Comment (hershpa) [reply docker/intel-dgpu-driver.Dockerfile:34]: Done!

File: docker/intel-dgpu-driver.Dockerfile (hunk 1 above)
Comment (hershpa) [reply docker/intel-dgpu-driver.Dockerfile]: This is new to me. :) Can you elaborate what I should try? Based on cp man page, --parents will use full source file name under DIRECTORY.

Comment (hershpa) [reply docker/intel-dgpu-driver.Dockerfile]: Would you like me to try this?
```
RUN for file in $(find /lib/modules/${KERNEL_FULL_VERSION}/updates/ -name "*.ko"); do \
    dname=$(dirname /opt$file); mkdir -p $dname; cp $file /opt --parents; done
```

Comment (hershpa) [reply docker/intel-dgpu-driver.Dockerfile]: Done!

**Review by uMartinXu — APPROVED** [review]
LGTM

**Pull request merged**